[FIX] account: don't try being smart when importing multiple files

qdp-odoo opened the pull request:
In case of multiple files being imported at the same time (for example, when an email with several attachments was received on a journal mail alias), the previous code was trying to detect if the given files were for the same invoice or not... But it was doing it wrongly and the processing order of attachment mattered, causing a diferent behaviour in one case or another. This bug just enlighted the fact we were trying too hard to do something smart here, where we'd actually don't need it.

So it has been decided to simplify things, and the algorithm will now simply create one invoice per valid attached file. Point. And if people have more complex needs, they should/can install the document app to handle them.

Use case:
* send an email to a mail alias with, attached in this specific order, a PDF (embedding an xml file) and a XLS.
* verify that odoo created 2 invoices, each linked to one of the files
* send an email to the same mail alias, with the same 2 files but attached on the mail in the inverse order: first the XLS and then the PDF
* odoo will create a single invoice, attaching the 2 files to it.

The reason is that
* the embedded XML is always done first
* in case 1 - PDF then XLS 

> Decoding of XML as invoice 1
> Decoding of PDF skipped because embedded XML was already imported as invoice 1 |Decoding of XLS as invoice 2 because the condition
> ```
>       # When receiving multiple files, if they have a different type, we supposed they are all linked
>       # to the same invoice.
>       if (
>           passed_file_data_list
>           and passed_file_data_list[-1]['filename'] != file_data['filename']
>           and passed_file_data_list[-1]['sort_weight'] != file_data['sort_weight']
>       ):
> ```
> 
>   is FALSE: passed_file_data_list[-1]['sort_weight'] == file_data['sort_weight'] == 100

* in case 2 - XLS then PDF 

> Decoding of XML as invoice 1
> Decoding of XLS as invoice 1 skipped because the same condition
> ```
>       # When receiving multiple files, if they have a different type, we supposed they are all linked
>       # to the same invoice.
>       if (
>           passed_file_data_list
>           and passed_file_data_list[-1]['filename'] != file_data['filename']
>           and passed_file_data_list[-1]['sort_weight'] != file_data['sort_weight']
>       ):
> ```
>  is now TRUE: passed_file_data_list[-1]['sort_weight'] == 20 and file_data['sort_weight'] == 100
> Decoding of PDF skipped because the embedded XML was already imported as invoice 1

ticket - 4510745


---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/203255.png)](https://mergebot.odoo.com/odoo/odoo/pull/203255)

aboo-odoo reviewed (commented):
@smetl , didn't we add these line to make sure separated XML and PDF are place in the same invoice ? I believe an example of use case was CFDI in Mexico where the XML is separated from the PDF but related to the same invoice.

aboo-odoo commented on addons/account/models/account_move.py:
```diff
@@ -3198,17 +3198,6 @@ def add_file_data_results(file_data, invoice):
-                add_file_data_results(file_data, invoices[-1])
-                close_file(file_data)
-                continue
-
```
I thought about sorting the list return by `_unwrap_edi_attachments` by weight, then by name but, would it be a solution: the same behavior will occur depending on whether you call you file a.pdf, b.xls and b.pdf, a.xls ?

qdp-odoo commented:
sorry guys, I ddin't know you'd check already the branch since the runbot was red... :shrug: 

I've solved the failing test, added some code to make an exception if we have both an XML and a PDF that includes the same XML attached, and added the corresponding test as well... Should be green, as it passes locally. :-)

qdp-odoo commented on addons/account/tests/test_account_incoming_supplier_invoice.py:
```diff
@@ -345,7 +345,14 @@ def test_extend_with_attachments_multi_pdf(self):
+        from odoo.tools.pdf import OdooPdfFileReader
+        with self.with_success_decoder() as decoded_files, \
+             patch.object(OdooPdfFileReader, 'getAttachments', return_value=[(xml1.name, '<test2></test2>')]), \
+             patch.object(type(self.env['account.move']), '_get_ubl_cii_builder_from_xml_tree', return_value=self.env['account.edi.xml.cii']):
```
neat! I didn't know... thanks Xavier :)

smetl commented on addons/account/tests/test_account_incoming_supplier_invoice.py:
```diff
@@ -345,7 +345,14 @@ def test_extend_with_attachments_multi_pdf(self):
+        with self.with_success_decoder() as decoded_files, \
+             patch.object(OdooPdfFileReader, 'getAttachments', return_value=[(xml1.name, '<test2></test2>')]), \
+             patch.object(self.registry['account.move'], '_get_ubl_cii_builder_from_xml_tree', return_value=self.env['account.edi.xml.cii']):
+            #the attachments are an XML and a PDF, that contains the same XML. In that case, we expect only one invoice to be created, with both files attached
```
You can remove all your diff because this use case is already managed. I tested by adding:
```py
        docx = self._create_dummy_docx_attachment()

        with self.with_success_decoder() as decoded_files:
            self._assert_extend_with_attachments({pdf1: 1, docx: 1}, origin='mail_alias')
            self.assertEqual(decoded_files, {pdf1.name})
        with self.with_success_decoder() as decoded_files:
            self._assert_extend_with_attachments({xml1: 1, pdf1: 1, docx: 1}, origin='mail_alias')
            self.assertEqual(decoded_files, {xml1.name})
```
and it works! There is a little mistake in `_unwrap_edi_attachments` that duplicates the `file_data` and messed up the import in that case: `break` is missing:

```py
    def _unwrap_edi_attachments(self):
        """Decodes ir.attachment and unwrap sub-attachment into a sorted list of
        dictionary each representing an attachment.

        :returns:           A list of dictionary for each attachment.
        * filename:         The name of the attachment.
        * content:          The content of the attachment.
        * type:             The type of the attachment.
        * xml_tree:         The tree of the xml if type is xml.
        * pdf_reader:       The pdf_reader if type is pdf.
        * attachment:       The associated ir.attachment if any
        * sort_weight:      The associated weigth used for sorting the arrays
        """
        to_process = []

        for attachment in self:
            supported_formats = attachment._get_edi_supported_formats()
            for supported_format in supported_formats:
                if supported_format['check'](attachment):
                    to_process += supported_format['decoder'](attachment.name, attachment.raw)
                    break  # <- HERE

        to_process.sort(key=lambda x: x['sort_weight'])

        return to_process
```

qdp-odoo replied:
> @[smetl](https://github.com/smetl) smetl [yesterday](https://github.com/odoo/odoo/pull/203255/files#r2020691202)
> You can remove all your diff because this use case is already managed. I tested by adding:
> 
>         docx = self._create_dummy_docx_attachment()
> 
>         with self.with_success_decoder() as decoded_files:
>             self._assert_extend_with_attachments({pdf1: 1, docx: 1}, origin='mail_alias')
>             self.assertEqual(decoded_files, {pdf1.name})
>         with self.with_success_decoder() as decoded_files:
>             self._assert_extend_with_attachments({xml1: 1, pdf1: 1, docx: 1}, origin='mail_alias')
>             self.assertEqual(decoded_files, {xml1.name})
> and it works! 
but it's not what we want (said @chklop): each valid file should create a new invoice. I added the exception for XML files that are embedded in PDF, but it's... wel...l an exception

smetl replied:
@qdp-odoo @chklop Then I disagree on this:
2 PDFs => 2 invoices
1 PDF + 1 sheet => 1 invoice
This is far more logical to me.

smetl commented:
@qdp-odoo I discussed it again with POs and we conclude PDF + docs => 1 invoice is fine.
=> https://github.com/odoo/odoo/pull/205328
